api/src/main/java/io/minio/MinioClient.java

```diff
@@ -1774,6 +1870,67 @@ public void copyObject(String bucketName, String objectName, String destBucketNa
     copyObject(bucketName, objectName, destBucketName, destObjectName, copyConditions, null);
   }
 
+  /**
+   * Copy a source object into a new object with the provided name in the provided bucket.
+   * optionally can take a key value CopyConditions as well for conditionally attempting
+   * copyObject.
+   *
+   * </p>
+   * <b>Example:</b><br>
+   *
+   * <pre>
+   * {@code minioClient.copyObject("my-bucketname", "my-objectname", sseSource, "my-destbucketname",
+   * "my-destobjname", copyConditions, sseTarget);}
+   * </pre>
+   *
+   * @param bucketName
+   *          Bucket name where the object to be copied exists.
+   * @param objectName
+   *          Object name source to be copied.
+   * @param destBucketName
+   *          Bucket name where the object will be copied to.
+   * @param destObjectName
+   *          Object name to be created, if not provided uses source object name
+   *          as destination object name.
+   * @param copyConditions
+   *          CopyConditions object with collection of supported CopyObject conditions.
+   * @param sseSource
+   *          Source Encryption metadata.
+   * @param sseTarget
+   *          Target Encryption metadata.
+   *
+   * @throws InvalidBucketNameException
+   *           upon an invalid bucket name, invalid object name.
+   * @throws NoSuchAlgorithmException
+   *           upon requested algorithm was not found during signature calculation
+   * @throws InvalidKeyException
+   *           upon an invalid access key or secret key
+   */
+  public void copyObject(String bucketName, String objectName, ServerSideEncryption sseSource, String destBucketName,
+                         String destObjectName, CopyConditions copyConditions, ServerSideEncryption sseTarget)
+      throws InvalidKeyException, InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException,
+      NoResponseException, ErrorResponseException, InternalException, IOException, XmlPullParserException,
+      InvalidArgumentException {
+
+    if ((sseTarget.getType() == ServerSideEncryption.Type.SSE_C) && (!this.baseUrl.isHttps())) {
+      throw new InvalidArgumentException("SSE_C operations must be performed over a secure connection.");
+    } else if ((sseTarget.getType() == (ServerSideEncryption.Type.SSE_KMS)) && (!this.baseUrl.isHttps())) {
+      throw new InvalidArgumentException("SSE_KMS operations must be performed over a secure connection.");
+    }
+    Map<String, String> headerTarget = new HashMap<>();
+    Map<String, String> headers = new HashMap<>();
+    sseTarget.marshal(headerTarget);
+    if (sseTarget.getType() == ServerSideEncryption.Type.SSE_C) {
+      Map<String, String> headerSource = new HashMap<>();
+      sseSource.marshal(headerSource);
+      headers.putAll(headerSource);
+      headers.putAll(headerTarget);
+    } else {
+      headers.putAll(headerTarget);
+    }
+    copyObject(bucketName, objectName, destBucketName, destObjectName, copyConditions, headers);
+  }
```

Comment:
this seems incorrect - the sse headers will not be sent here if source is encrypted with sse-c but not target
```
    Map<String, String> headerTarget = new HashMap<>();
    Map<String, String> headerSource = new HashMap<>();
    Map<String, String> headers = new HashMap<>();
    sseTarget.marshal(headerTarget);
    sseSource.marshal(headerSource);
    headers.putAll(headerSource);
    headers.putAll(headerTarget);
    copyObject(bucketName, objectName, destBucketName, destObjectName, copyConditions, headers);
  }
```

Reply:
Okay, I will check it.